src/lib/template-factory.ts

```diff
@@ -44,19 +44,40 @@ export type TemplateFactory = (result: TemplateResult) => Template;
 export function templateFactory(result: TemplateResult) {
   let templateCache = templateCaches.get(result.type);
   if (templateCache === undefined) {
-    templateCache = new Map<TemplateStringsArray, Template>();
+    templateCache = new Map<string, Template>();
     templateCaches.set(result.type, templateCache);
   }
-  let template = templateCache.get(result.strings);
+  let key = templateKeys.get(result.strings);
+  if (key !== undefined) {
+    // If templateKeys has the key it's guaranteed to exist in the templateCache
+    return templateCache.get(key)!;
+  }
+
+  // If the result.strings are new, generate a key and set it in templateKeys
+  key = result.strings.join(marker);
+  templateKeys.set(result.strings, key);
+
+  let template = templateCache.get(key);
   if (template === undefined) {
     template = new Template(result, result.getTemplateElement());
-    templateCache.set(result.strings, template);
+  } else {
+    // If a template already exists, refresh the key object in the templateCache
+    templateCache.delete(key);
   }
+  templateCache.set(key, template);
```

Comment:
We only need to delete/reset when we generate a new key. I think it's fine to add a variable to track that.

Reply:
This set only happens when a new key is generated. There is an early return up above that happens when the key exists. It was originally structured like an if-else tree, but most of the code in this project seems to prefer early returns so I followed that style.

Reply:
I could add some more comments to explain the routine if necessary. I was worried I'd add too many comments if I explained it more thoroughly so I kept the comments to what I thought was the minimum necessary